src/server/campaignd/server.cpp

```diff
@@ -898,30 +1065,113 @@ void server::handle_upload(const server::request& req)
 		}
 
 		const std::string& filename = (*campaign)["filename"].str();
-		data["title"] = (*campaign)["title"];
 		data["name"] = "";
-		data["campaign_name"] = (*campaign)["name"];
-		data["author"] = (*campaign)["author"];
-		data["description"] = (*campaign)["description"];
-		data["version"] = (*campaign)["version"];
-		data["timestamp"] = (*campaign)["timestamp"];
-		data["original_timestamp"] = (*campaign)["original_timestamp"];
-		data["icon"] = (*campaign)["icon"];
-		data["type"] = (*campaign)["type"];
-		data["tags"] = (*campaign)["tags"];
 		find_translations(data, *campaign);
 
 		add_license(data);
+		
+		const std::string& new_version = (*campaign)["version"].str();
+		//#TODO: probably make hash from new_version + required_wesnoth_version ?
+		const std::string& file_hash = utils::md5(new_version).hex_digest();
+
+		//sorted version map
+		auto version_map = get_version_map(*campaign);
 
+		//#TODO: add gfgtdf's stuff about required_wesnoth_version here
+		config version_cfg = config("version", new_version);
+		version_cfg["filename"] = "/full_pack_" + file_hash + ".gz";
+
+		version_map.erase(version_info(new_version));
+		version_map.emplace(version_info(new_version), version_cfg);
+		(*campaign).remove_children("version", [&new_version](const config& old_cfg) 
+			{
+				return old_cfg["version"].str() == new_version; 
+			}
+		);
+		(*campaign).add_child("version", version_cfg);
+
+		//Let's write the full_pack file
 		{
-			filesystem::atomic_commit campaign_file(filename);
+			filesystem::atomic_commit campaign_file(filename + version_cfg["filename"].str());
 			config_writer writer(*campaign_file.ostream(), true, compress_level_);
 			writer.write(data);
 			campaign_file.commit();
 		}
 
-		(*campaign)["size"] = filesystem::file_size(filename);
+		(*campaign)["size"] = filesystem::file_size(filename + version_cfg["filename"].str()) + filesystem::file_size(filename + "/addon.cfg");
```

Comment:
There's basically no point to this change since the `size` attribute is primarily used to inform the user of the add-on download size and the size of the metadata is both negligible and irrelevant, since the client receives the full metadata in the `[request_campaign_list]` response instead.

This should've been in a separate commit since it's both orthogonal to the actual commit ("Split addons meta from the server config") and completely changes existing behaviour.